src/kbmod/search/raw_image.h

```diff
@@ -35,88 +40,98 @@ class RawImage {
     unsigned get_width() const { return width; }
     unsigned get_height() const { return height; }
     unsigned get_npixels() const { return width * height; }
+    const Image& get_image() const { return image; }
+    Image& get_image() { return image; }
+    void set_image(Image& other) { image = other; }
 
-    // Inline pixel functions.
-    float get_pixel(int x, int y) const {
-        return (x >= 0 && x < width && y >= 0 && y < height) ? pixels[y * width + x] : NO_DATA;
+    inline bool contains(const Index& idx) const {
+        return idx.i >= 0 && idx.i < width && idx.j >= 0 && idx.j < height;
     }
 
-    bool pixel_has_data(int x, int y) const {
-        return (x >= 0 && x < width && y >= 0 && y < height) ? pixels[y * width + x] != NO_DATA : false;
-    }
+    inline bool contains(const Point& p) const { return p.x >= 0 && p.y < width && p.y >= 0 && p.y < height; }
 
-    void set_pixel(int x, int y, float value) {
-        if (x >= 0 && x < width && y >= 0 && y < height) pixels[y * width + x] = value;
-    }
-    const std::vector<float>& get_pixels() const { return pixels; }
-    float* data() { return pixels.data(); }  // Get pointer to pixels
+    inline float get_pixel(const Index& idx) const { return contains(idx) ? image(idx.j, idx.i) : NO_DATA; }
 
-    // Get the interpolated brightness of a real values point
-    // using the four neighboring pixels.
-    float get_pixel_interp(float x, float y) const;
+    bool pixel_has_data(const Index& idx) const { return get_pixel(idx) != NO_DATA ? true : false; }
 
-    // Check if two raw images are approximately equal.
-    bool approx_equal(const RawImage& imgB, float atol) const;
+    void set_pixel(const Index& idx, float value) {
+        // we should probably be letting Eigen freak out about setting an impossible
+        // index instead of silently just nod doing it; but this is how it is
```

Comment:
Does Eigen raise an error for an impossible index? The problem we were having with apply_mask makes me worry that it doesn't.

Regardless, I agree that this shouldn't be silent. Maybe something like:

```
        if (!contains(idx))
            throw std::runtime_error("Invalid index");
        image(idx.j, idx.i) = value;
```

Reply:
Eigen will complain when accessing out of bounds.See [manual](https://eigen.tuxfamily.org/dox/group__QuickRefPage.html). It's possible to avoid range checking by defining NDEBUG (in which case asserts are skipped too) or by using `.coeff` or `.coeffRef` accessors.

It's `set_pixel` that never raises and `get_pixel` returns `NO_DATA` value when out of bounds. We never know if we're outside of the array or not. In fact we count on it in some of the functions.

Raising an error now.